presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java

```diff
@@ -0,0 +1,471 @@
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable cause)
+                {
+                    try (SetThreadName ignored = new SetThreadName("Task-%s", taskId)) {
+                        taskStateMachine.failed(cause);
+
+                        // record driver is finished
+                        remainingDrivers.decrementAndGet();
+
+                        // fire failed event with cause
+                        splitMonitor.splitFailedEvent(taskId, getDriverStats(), cause);
+                    }
+                }
+
+                private DriverStats getDriverStats()
+                {
+                    DriverContext driverContext = splitRunner.getDriverContext();
+                    DriverStats driverStats;
+                    if (driverContext != null) {
+                        driverStats = driverContext.getDriverStats();
+                    }
+                    else {
+                        // split runner did not start successfully
+                        driverStats = new DriverStats();
+                    }
+
+                    return driverStats;
+                }
+            }, notificationExecutor);
+        }
+    }
+
+    private synchronized void checkTaskCompletion()
+    {
+        if (taskStateMachine.getState().isDone()) {
+            return;
+        }
+
+        // are there more partition splits expected?
+        for (DriverSplitRunnerFactory driverSplitRunnerFactory : driverRunnerFactoriesWithSplitLifeCycle.values()) {
+            if (!driverSplitRunnerFactory.isNoMoreDriverRunner()) {
+                return;
+            }
+        }
+        // do we still have running tasks?
+        if (remainingDrivers.get() != 0) {
+            return;
+        }
+
+        // Cool! All done!
+        taskStateMachine.finished();
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("taskId", taskId)
+                .add("remainingDrivers", remainingDrivers.get())
+                .toString();
+    }
+
+    private class DriverSplitRunnerFactory
+    {
+        private final DriverFactory driverFactory;
+        private final PipelineContext pipelineContext;
+
+        private final AtomicInteger pendingCreation = new AtomicInteger();
+        private final AtomicBoolean noMoreDriverRunner = new AtomicBoolean();
+        private boolean closed;
+
+        private DriverSplitRunnerFactory(DriverFactory driverFactory, boolean partitioned)
+        {
+            this.driverFactory = requireNonNull(driverFactory, "driverFactory is null");
+            this.pipelineContext = taskContext.addPipelineContext(driverFactory.getPipelineId(), driverFactory.isInputDriver(), driverFactory.isOutputDriver(), partitioned);
+        }
+
+        public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitionedSplit)
```

Comment:
nit: Why this method and `createDriver` has to be public? :)

Reply:
That's usually a convention we are trying to maintain in the codebase. The encapsulation of the inner class is enforced by the `private` modifier on the class level. `public` modifiers on it's method usually indicate that the method is called from outside the inner class by the outer class. Too bad the Java compiler doesn't enforce it, and allows private inner class methods to be called from the outer class breaking the encapsulation.